### PR TITLE
Add converter to convert v1 model checkpoint to v2 and an example notebook to use converter

### PR DESCRIPTION
**Pull request by hwpang**
## Description
@am2145 and I add a `v1_to_v2.py` file containing the utility functions to map a v1 checkpoint dictionary to a v2 checkpoint dictionary format. It converts the information from v1 arguments and v1 state dict to v2 hyperparameters and v2 state dict appropriately. As suggested by @KnathanM, we also add a jupyter notebook where the user can examine the contents of the checkpoint dictionary, convert the model, and load the converted model.

We test the converter using a regression model trained in v1. We compare the prediction made by the v1 model using the v1 code to the prediction made by the converted model using the v2 code, and the predictions are the same down to the 6th decimal place.

### CODE REVIEW

**Comment by davidegraff** [comment]
great work and thanks for taking care of this!
- let's not put example notebooks in the core module. We don't want to distribute these in the source wheel, so we should move these to the top level `CHEMPROP_DIR/examples`
- I think our users would be well-served by making this into a CLI utility. I'm thinking somethin like so:
```
$ chemprop convert --help
usage: chemprop convert input_path [output_path]

positional arguments:
  input_path     the path of a saved v1 model

optional arguments:
  output_path  the path to which the converted model should be saved. If not provided, appends '_v2' to `input_path` and saves the new model to that path.

**Comment by hwpang** [comment]
@davidegraff Thanks for the suggestions! I moved the notebook to the top level, and I will work on the CLI utility.

**Comment by hwpang** [comment]
In the new commit I add the CLI for the model converter. I made both the `input_path` and `output_path` keyword arguments and set `input_path` to be required after discussing this with @kevingreenman.

File: chemprop/v2/cli/convert.py [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+
+    @classmethod
+    def func(cls, args: Namespace):
+        args = process_args(args)
+        main(args)
+
+def add_convert_args(parser: ArgumentParser) -> ArgumentParser:
+    parser.add_argument(
+        "-i",
+        "--input_path",
```
Comment (KnathanM) [inline chemprop/v2/cli/convert.py:26]: It could be good to keep to our v2 convention to use dashes for long args `--input-path`.

File: chemprop/v2/cli/convert.py [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+        type=str,
+        required=True,
+        help="Path to a v1 model .pt checkpoint file.",
+    )
+    parser.add_argument(
+        "-o",
+        "--output_path",
+        type=str,
+        help="Path at which the converted model will be saved.",
```
Comment (KnathanM) [inline chemprop/v2/cli/convert.py:35]: A note about what the default is would be helpful in the help string. Something like "If not specified, defaults to the name of the input file suffixed with '_v2.ckpt' in your current working directory."

File: chemprop/v2/cli/convert.py [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+    )
+    return parser
+
+def process_args(args: Namespace) -> Namespace:
+    args.input_path = Path(args.input_path)
+    args.output_path = Path(args.output_path or Path.cwd() / (args.input_path.stem + "_v2.ckpt"))
+    return args
+
+def main(args: Namespace) -> None:
+    logging.info(f"Converting v1 model checkpoint '{args.input_path}' to v2 model checkpoint '{args.output_path}'...")
```
Comment (KnathanM) [inline chemprop/v2/cli/convert.py:45]: I believe to use logging you need this line after your imports at the top of the script `logger = logging.getLogger(__name__)` and then use `logger.info(...)` instead of `logging.info(...)`.

File: chemprop/v2/cli/convert.py [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+    convert_model_v1_to_v2_file(str(args.input_path), str(args.output_path))
+
+if __name__ == "__main__":
+    parser = ArgumentParser()
+    ConvertSubcommand.add_args(parser)
+
+    logging.basicConfig(stream=sys.stdout, level=logging.DEBUG, force=True)
```
Comment (KnathanM) [inline chemprop/v2/cli/convert.py:52]: Same as above

**Review by KnathanM — CHANGES_REQUESTED** [review]

**Comment by hwpang** [comment]
@KnathanM Thanks for your comment! I made the changes as suggested.

File: chemprop/v2/cli/convert.py [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+        required=True,
+        help="Path to a v1 model .pt checkpoint file.",
+    )
+    parser.add_argument(
+        "-o",
+        "--output-path",
+        type=str,
+        help="Path at which the converted model will be saved. If not specified, it defaults to the name of the input file suffixed with '_v2.ckpt' in your current working directory.",
+    )
+    return parser
```
Comment (davidegraff) [inline chemprop/v2/cli/convert.py:39]: can you just move this all into `ConvertSubcommand.add_args`?

File: chemprop/v2/cli/convert.py [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+        main(args)
+
+def add_convert_args(parser: ArgumentParser) -> ArgumentParser:
+    parser.add_argument(
+        "-i",
+        "--input-path",
+        type=str,
+        required=True,
+        help="Path to a v1 model .pt checkpoint file.",
+    )
```
Comment (davidegraff) [inline chemprop/v2/cli/convert.py:32]:
```suggestion
    parser.add_argument(
        "input_path",
        help="Path to a v1 model .pt checkpoint file.",
    )
```
This makes `input-path` into a positional arg, which is always required. IMO it makes sense that a user calls this with only two arguments: (1) input path and (2, optional) output path

File: chemprop/v2/cli/convert.py [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+    parser.add_argument(
+        "-o",
+        "--output-path",
+        type=str,
```
Comment (davidegraff) [inline chemprop/v2/cli/convert.py:36]:
```suggestion
        "output_path",
        nargs="?",
```
This makes `output_path` into an optional, positional argument (after `input_path`).

File: chemprop/v2/cli/convert.py [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+    return parser
+
+def process_args(args: Namespace) -> Namespace:
+    args.input_path = Path(args.input_path)
+    args.output_path = Path(args.output_path or Path.cwd() / (args.input_path.stem + "_v2.ckpt"))
+    return args
+
+def main(args: Namespace) -> None:
+    logging.info(f"Converting v1 model checkpoint '{args.input_path}' to v2 model checkpoint '{args.output_path}'...")
+    convert_model_v1_to_v2_file(str(args.input_path), str(args.output_path))
```
Comment (davidegraff) [inline chemprop/v2/cli/convert.py:48]: move all this code into `ConvertSubcommand.func`

File: chemprop/v2/utils/v1_to_v2.py [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+    state_dict_v2["message_passing.W_o.bias"] = state_dict_v1[
+        "encoder.encoder.0.W_o.bias"
+    ]
+
+    state_dict_v2["readout.loc"] = torch.from_numpy(
+        model_v1["data_scaler"]["means"]
+    ).unsqueeze(0)
+    state_dict_v2["readout.scale"] = torch.from_numpy(
+        model_v1["data_scaler"]["stds"]
+    ).unsqueeze(0)
```
Comment (davidegraff) [inline chemprop/v2/utils/v1_to_v2.py:41]: only `RegressionReadout` classes have the `loc` and `scale` attributes right?

File: chemprop/v2/utils/v1_to_v2.py [hunk 10]
```diff
@@ -0,0 +1,140 @@
...
+
+def convert_hyper_parameters_v1_to_v2(model_v1: dict) -> dict:
+    """Converts v1 model dictionary to v2 hyper_parameters dictionary
+    """
+    hyper_parameters_v2 = {}
+
+    args_v1 = model_v1["args"]
+    hyper_parameters_v2[
+        "batch_norm"
+    ] = False  # It doesn't seem that batch norm is done in v1
```
Comment (davidegraff) [inline chemprop/v2/utils/v1_to_v2.py:62]: yes, batch norm is new in v2

File: chemprop/v2/utils/v1_to_v2.py [hunk 11]
```diff
@@ -0,0 +1,140 @@
...
+    model_v2 = {}
+    
+    model_v2["epoch"] = None  # Not defined in v1 model dictionary
+    model_v2["global_step"] = None  # Not defined in v1 model dictionary
+    model_v2["pytorch-lightning_version"] = __version__ 
+    model_v2["state_dict"] = convert_state_dict_v1_to_v2(model_v1)
+    model_v2["loops"] = None  # Not defined in v1 model dictionary
+    model_v2["callbacks"] = None  # Not defined in v1 model dictionary
+    model_v2["optimizer_states"] = None  # Not defined in v1 model dictionary
+    model_v2["lr_schedulers"] = None  # Not defined in v1 model dictionary
```
Comment (davidegraff) [inline chemprop/v2/utils/v1_to_v2.py:129]: remove comments

File: chemprop/v2/cli/convert.py [hunk 12]
```diff
@@ -0,0 +1,58 @@
...
+
+if __name__ == "__main__":
+    parser = ArgumentParser()
+    ConvertSubcommand.add_args(parser)
+
+    logger.basicConfig(stream=sys.stdout, level=logging.DEBUG, force=True)
```
Comment (KnathanM) [inline chemprop/v2/cli/convert.py:54]: Sorry Hao-Wei, I should have been more clear. The logging syntax is
`logging.basicConfig(...)`
`logger.info(...)`
These need to be changed on lines 47 and 54.

File: chemprop/v2/cli/convert.py (hunk 4 above)
Comment (KnathanM) [reply chemprop/v2/cli/convert.py]: I was wrong about this one. It is `logging.basicConfig(...)`

File: chemprop/v2/utils/v1_to_v2.py [hunk 13]
```diff
@@ -0,0 +1,140 @@
...
+    model_v2["hparams_name"] = "kwargs"
+    model_v2["hyper_parameters"] = convert_hyper_parameters_v1_to_v2(model_v1)
+
+    return model_v2
+
+def convert_model_v1_to_v2_file(model_v1_file: str, model_v2_file: str) -> None:
+    """Converts a v1 model .pt file to a v2 model .ckpt file
+    """
+    model_v1 = torch.load(model_v1_file)
+    model_v2 = convert_model_v1_to_v2(model_v1)
```
Comment (davidegraff) [inline chemprop/v2/utils/v1_to_v2.py:139]: can we make the variable names more clear? `model_v*` makes it sound like these variables are actual models/DMPNNs, but these are just dictionaries containing model (hyper)parameters?

File: chemprop/v2/cli/convert.py [hunk 14]
```diff
@@ -0,0 +1,58 @@
...
+        "-o",
+        "--output-path",
+        type=str,
+        help="Path at which the converted model will be saved. If not specified, it defaults to the name of the input file suffixed with '_v2.ckpt' in your current working directory.",
+    )
+    return parser
+
+def process_args(args: Namespace) -> Namespace:
+    args.input_path = Path(args.input_path)
+    args.output_path = Path(args.output_path or Path.cwd() / (args.input_path.stem + "_v2.ckpt"))
```
Comment (davidegraff) [inline chemprop/v2/cli/convert.py:43]: I think we want the default output path to be like so: `input_path = path/to/model.pt` → `output_path = path/to/model_v2.ckpt`. Right now it will write to `$CWD/model_v2.ckpt`

**Review by davidegraff — COMMENTED** [review]
a bunch of tiny comments, but happy to approve after they're addressed

File: chemprop/v2/cli/convert.py [hunk 15]
```diff
@@ -0,0 +1,44 @@
...
+    @classmethod
+    def add_args(cls, parser: ArgumentParser) -> ArgumentParser:
+        parser.add_argument(
+            "input_path",
+            help="Path to a v1 model .pt checkpoint file.",
+        )
+        parser.add_argument(
+            "output_path",
+            nargs="?",
+            help="Path at which the converted model will be saved. If not specified, it defaults to the name of the input file suffixed with '_v2.ckpt' in your current working directory.",
```
Comment (KnathanM) [inline chemprop/v2/cli/convert.py:25]: We changed the default:
"If not specified, it defaults to the name of the input file suffixed with '_v2.ckpt' in the same parent directory."

I think how I worded this earlier could be confusing. Maybe a simple example is more clear:
"If not specified and the input file is `/path/to/checkpoint/checkpoint.pt`, the output will default to `/path/to/checkpoint/checkpoint_v2.ckpt`. 

Please choose what sounds clearer to you.

File: chemprop/v2/cli/convert.py [hunk 16]
```diff
@@ -0,0 +1,44 @@
...
+        )
+        return parser
+
+    @classmethod
+    def func(cls, args: Namespace):
+        args.input_path = Path(args.input_path)
+        args.output_path = Path(args.output_path or args.input_path.parent / (args.input_path.stem + "_v2.ckpt"))
+
+        logger.info(f"Converting v1 model checkpoint '{args.input_path}' to v2 model checkpoint '{args.output_path}'...")
+        convert_model_file_v1_to_v2(str(args.input_path), str(args.output_path))
```
Comment (KnathanM) [inline chemprop/v2/cli/convert.py:35]: `args.input_path` starts as a string. You then convert it to a path using `Path()`. The on line 35 you convert it back to a string. Would it be okay to just leave `args.input_path` and `args,output_path` as strings?

File: chemprop/v2/utils/v1_to_v2.py (hunk 10 above)
Comment (KnathanM) [reply chemprop/v2/utils/v1_to_v2.py:64]: I think you could remove the comment or change it to " # Batch norm is new to v2". My preference is remove.
Also I don't know why this hyperparameter is spread over three lines. I think it could be collapsed to just one.
`hyper_parameters_v2["batch_norm"] = False`

**Comment by hwpang** [comment]
@KnathanM @davidegraff Thank you so much for the comments! In the most recent push, I have addressed all the comments and cleaned up the commits. I also add a unit test for the converter to test that we can load the converted model.

**Comment by hwpang** [comment]
@KnathanM I just saw your newer comments. I will work on them now.

File: chemprop/v2/utils/v1_to_v2.py (hunk 10 above)
Comment (hwpang) [reply chemprop/v2/utils/v1_to_v2.py]: Thanks! I remove the comment in the new push. I use `black` as my default formatter, and it tries to format the code if the code is too long to improve readability. In the new push I also make this into one line.

**Comment by hwpang** [comment]
@KnathanM I address your comments in the new push! I think this PR is ready for review and merge.

File: chemprop/v2/cli/convert.py [hunk 17]
```diff
@@ -0,0 +1,44 @@
...
+    @classmethod
+    def add_args(cls, parser: ArgumentParser) -> ArgumentParser:
+        parser.add_argument(
+            "input_path",
+            help="Path to a v1 model .pt checkpoint file.",
+        )
+        parser.add_argument(
+            "output_path",
+            nargs="?",
+            help="Path at which the converted model will be saved. If not specified and the input file is /path/to/checkpoint/model.pt, the output will default to /path/to/checkpoint/model_v2.ckpt",
```
Comment (davidegraff) [inline chemprop/v2/cli/convert.py:25]:
```suggestion
            help="The path to which the converted model will be saved. If not specified and the input file is '/path/to/checkpoint/model.pt', the output will default to '/path/to/checkpoint/model_v2.ckpt'",
```

File: chemprop/v2/tests/unit/utils/test_converter.py [hunk 18]
```diff
@@ -0,0 +1,20 @@
...
+TEST_DIR = Path(__file__).parents[2]
+TEST_DATA_DIR = TEST_DIR / "data"
+
+
+@pytest.fixture
+def example_model_v1_path():
+    return TEST_DATA_DIR / "example_model_v1.pt"
+
+def test_converter(example_model_v1_path):
+    convert_model_file_v1_to_v2(example_model_v1_path, "example_model_v2.ckpt")
```
Comment (davidegraff) [inline chemprop/v2/tests/unit/utils/test_converter.py:17]: use the pytest `tmp_path` [fixture](https://docs.pytest.org/en/6.2.x/tmpdir.html) here so that you don't actually need to clean up the written checkpoint. The reason this matters is that this test, as written, could fail for file writing/removal, which is outside the scope of the test.

**Review by davidegraff — COMMENTED** [review]
Broad question, are there other ways we can test this utility? I can't think of any right now, but simply checking that we indeed write a file seems a little light. No worries if not and happy to accept after the changes, but something to keep in mind.

File: chemprop/v2/cli/convert.py (hunk 16 above)
Comment (davidegraff) [reply chemprop/v2/cli/convert.py:35]: can the converter function not accept `Path`s?

**Comment by davidegraff** [comment]
Also, you're just running `black` with the default settings right (i.e., not supply any additional flags)? This ensures that you use the default settings from `pyproject.toml`

**Comment by hwpang** [comment]
@davidegraff Thanks for the comments! In the new push, I (1) update the unit test to test that the converted model can make predictions sufficiently close to the original model and (2) use `tmp_path` to create a temporary path for the checkpoint file made by the converter.

**Review by davidegraff — APPROVED** [review]
👍  good work, thanks!

File: chemprop/v2/cli/convert.py (hunk 16 above)
Comment (hwpang) [reply chemprop/v2/cli/convert.py]: I thought it couldn't, but I just tested and it seemed to work fine with `Paths`, so I changed this part of code to pass just `args.input_path` and `args.output_path` in without converting them into `str`.

**Review by KnathanM — APPROVED** [review]
